plutus-core/testlib/Test/Tasty/Extras.hs

```diff
@@ -50,31 +65,116 @@ ghcVersion = showVersion compilerVersion
+    (<>) = (*>)
 
--- | Run a 'TestTree' of tests with an empty prefix.  This doesn't actually run
--- the tests: instead it runs a computation in the Reader monad.
-runTestNested :: TestNested -> TestTree
-runTestNested = runTestNestedIn []
+instance unit ~ () => Monoid (Layer a unit) where
+    mempty = pure ()
 
--- | Descend into a name prefix.
-testNested :: FilePath -> [TestNested] -> TestNested
-testNested folderName =
-    local (++ [folderName]) . fmap (testGroup folderName) . sequence
+instance unit ~ () => IsList (Layer a unit) where
+    type Item (Layer a unit) = a
+    fromList = traverse_ embed
+    toList layer = runF (unLayer layer) mempty $ uncurry (:)
+
+-- | Embed the given value into a 'Layer'-like type (either 'Layer' itself or a monad transformer
+-- stack with 'Layer' at the bottom).
+embed :: MonadFree ((,) a) m => a -> m ()
+embed x = liftF (x, ())
 
--- | Like `testNested` but adds a subdirectory corresponding to the GHC version being used.
-testNestedGhc :: FilePath -> [TestNested] -> TestNested
-testNestedGhc folderName = testNested (folderName </> ghcVersion)
+-- | Collapse the given 'Layer' into a single element by converting it into a list, applying the
+-- given function to the result and 'embed'ding it back.
+nestWith :: ([a] -> a) -> Layer a () -> Layer a ()
+nestWith f = embed . f . toList
+
+newtype TestNestedM r = TestNestedM
+    { unTestNestedM :: ReaderT [FilePath] (Layer TestTree) r
+    } deriving newtype
+        (Functor, Applicative, Monad, MonadReader [FilePath], MonadFree ((,) TestTree))
+
+-- | A 'TestTree' of tests under some name prefix.
+type TestNested = TestNestedM ()
+
+instance unit ~ () => Semigroup (TestNestedM unit) where
+    (<>) = (*>)
+
+instance unit ~ () => Monoid (TestNestedM unit) where
+    mempty = pure ()
+
+-- | Run a 'TestTree' of tests with a given name prefix. This doesn't actually run the tests:
+-- instead it runs a computation in the Reader monad.
+runTestNestedM :: [String] -> TestNested -> TestTree
+runTestNestedM []   _    = error "Path cannot be empty"
+runTestNestedM path test = testGroup (last path) . toList $ runReaderT (unTestNestedM test) path
+
+-- | Descend into a folder.
+testNestedNamedM
+    :: FilePath    -- ^ The folder.
+    -> String      -- ^ The name of the test to render in CLI.
+    -> TestNested
+    -> TestNested
+testNestedNamedM folderName testName
+    = TestNestedM
+    . local (++ [folderName])
+    . mapReaderT (nestWith $ testGroup testName)
+    . unTestNestedM
+
+-- | Descend into a folder.
+testNestedM :: FilePath -> TestNested -> TestNested
+testNestedM folderName = testNestedNamedM folderName folderName
+
+-- | Like 'testNestedM' but adds a subdirectory corresponding to the GHC version being used.
+testNestedGhcM :: TestNested -> TestNested
+testNestedGhcM = testNestedM ghcVersion
+
+runTestNested :: [String] -> [TestNested] -> TestTree
+runTestNested path = runTestNestedM path . fold
```

Comment:
These are really trivial with just `. fold` in the end, but I've tried refactoring all of the tests to use the `do`-notation and in the end realized it wasn't worth it, so I just replicated the old API such that nothing much needs to change with the tests.